### PR TITLE
DBZ-6591: Document read preference changes in 2.4

### PR DESCRIPTION
**Pull request by jcechace**
https://issues.redhat.com/browse/DBZ-6591

### CODE REVIEW

**Comment by jcechace** [comment]
@roldanbob feel free to push or suggest any changes as you see fit.

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 1]
```diff
@@ -95,6 +95,13 @@ In this way the connector dynamically adjusts to changes in replica set membersh
 * link:https://docs.mongodb.com/manual/tutorial/remove-shards-from-cluster/[Shard removal]
 * link:https://docs.mongodb.com/manual/changeStreams/[Change Streams]
 
+// Type: concept
+// Title: Description of how the MongoDB connector uses read preference
+// ModuleID: description-of-how-the-mongodb-connector-uses-read-preference
+[id="read-preference"]
+=== Read Preference
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:102]:
```suggestion
// Title: Description of how the MongoDB connector uses the MongoDB read preference
// ModuleID: description-of-how-the-mongodb-connector-uses-the-mongodb-read-preference
[id="read-preference"]
=== Read Preference
```

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 2]
```diff
@@ -95,6 +95,13 @@ In this way the connector dynamically adjusts to changes in replica set membersh
...
+Starting with Debezium 2.4 the MongoDB connector honors the link:https://www.mongodb.com/docs/manual/core/read-preference/[read preference] configured in either xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`] or xref:mongodb-property-mongodb-connection-string-shard-params[`mongodb.connection.string.shard.params`] for sharded clusters when xref:mongodb-property-mongodb-connection-mode[`mongodb.connection.mode`] is set to `replica_set`.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]:
```suggestion
Beginning with Debezium 2.4, when the MongoDB connector captures changes from a sharded cluster, and the xref:mongodb-property-mongodb-connection-mode[`mongodb.connection.mode`] is set to `replica_set`, the connector honors the link:https://www.mongodb.com/docs/manual/core/read-preference/[read preference] specified in either the xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`] or the xref:mongodb-property-mongodb-connection-string-shard-params[`mongodb.connection.string.shard.params`].
```

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 3]
```diff
@@ -1371,14 +1378,23 @@ Connectors that capture changes from a sharded MongoDB cluster use this connecti
 After the initial discovery process, connection strings are generated for each individual shard.
 ====
 
+|[[mongodb-property-mongodb-connection-string-shard-params]]<<mongodb-property-mongodb-connection-string-shard-params, `+mongodb.connection.string.shard.params+`>>
+|No default
+|Specifies the URL parameters of https://www.mongodb.com/docs/manual/reference/connection-string/[connection string] that the connector uses to connect to individual shards of MongoDB sharded cluster.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:1383]:
```suggestion
|Specifies the URL parameters of the https://www.mongodb.com/docs/manual/reference/connection-string/[connection string] that the connector uses to connect to individual shards of a MongoDB sharded cluster.
```
Does this property accept multiple strings? If so, I'd use the plural to reflect that:

>Specifies the URL parameters of the https://www.mongodb.com/docs/manual/reference/connection-string/[connection string**s**] that the connector uses to connect to individual shards of a MongoDB sharded cluster.`

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 4]
```diff
@@ -1371,14 +1378,23 @@ Connectors that capture changes from a sharded MongoDB cluster use this connecti
...
+
+[NOTE]
+====
+This property only takes effect when xref:mongodb-property-mongodb-connection-mode[`mongodb.connection.mode`] is set to `replica_set`.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:1387]:
```suggestion
This property applies only when the xref:mongodb-property-mongodb-connection-mode[`mongodb.connection.mode`] is set to `replica_set`.
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
As always, let me know if you have any questions, @jcechace

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc (hunk 2 above)
Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: @roldanbob  I believe this isn't right. Your wording changes the meaning. The important note here is that `mongodb.connection.string` is used for everything but sharded cluster  with `mongodb.connection.mode` set to `replica_set`. In such case `mongodb.connection.string.shard.params` is used instead

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc (hunk 3 above)
Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:1383]: No it doesn't. It's a single query string. e.g. `readPreference=primary&sslEnabled=false`

Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:1383]: Although technically there are multiple connection strings  constructed with this parameter so I'm not quite sure whether plural or singular ought to be used here.  WDYT?

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc (hunk 2 above)
Comment (roldanbob) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: Thanks, @jcechace I think that I've addressed the issue. Please take a look to verify.

Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: @roldanbob  not quite... though reading my original post it is ambiguous.

So.. By default in 2.4 the connector honors the setting in `mongodb.connection.string`. However specifically (and only) for sharded clusters with `mongodb.connection.mode` set to `replica_set` the setting is taken from `mongodb.connection.string.shard.params`.

This is because for all other combinations of connnector config and MongoDB topology the connector uses the connection String set in  `mognodb.connection.string` to connect to the database AND capture changes. However, for sharded clusters with `mongodb.connection.mode` set to `replica_set` (as opposed to `shared`) the initial connection string from `mongodb.connection.string` is used ONLY to read the cluster topology from the `config.shards` collection. Once that is done a new connection string is constructed for each shard by combining the information retrieved from `config.shards` collection and the parameters specified in `mongodb.connection.string.shard.pamars`.

Comment (roldanbob) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: Ok, yeah, thanks for teasing that out a little more.  I was/am still a little confused about the linkage between how Debezium connects to a shard, and the read preference mode, but it's beginning to make sense. I guess that the read preference is implied from the connection interface? 
 I'm not sure that I've cracked this nut yet, but here's another go at it. Getting closer?

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 5]
```diff
@@ -1371,14 +1378,23 @@ Connectors that capture changes from a sharded MongoDB cluster use this connecti
 After the initial discovery process, connection strings are generated for each individual shard.
 ====
 
+|[[mongodb-property-mongodb-connection-string-shard-params]]<<mongodb-property-mongodb-connection-string-shard-params, `+mongodb.connection.string.shard.params+`>>
+|No default
+|Specifies the URL parameters of the https://www.mongodb.com/docs/manual/reference/connection-string/[connection string] that the connector uses to connect to individual shards of a MongoDB sharded cluster.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:1383]:
```suggestion
|Specifies the URL parameters of the https://www.mongodb.com/docs/manual/reference/connection-string/[connection string], including read preferences, that the connector uses to connect to individual shards of a MongoDB sharded cluster.
```

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 6]
```diff
@@ -95,6 +95,13 @@ In this way the connector dynamically adjusts to changes in replica set membersh
 * link:https://docs.mongodb.com/manual/tutorial/remove-shards-from-cluster/[Shard removal]
 * link:https://docs.mongodb.com/manual/changeStreams/[Change Streams]
 
+// Type: concept
+// Title: Description of how the MongoDB connector uses the MongoDB read preference
+// ModuleID: description-of-how-the-mongodb-connector-uses-the-mongodb-read-preference
+[id="read-preference"]
+=== Read Preference
+Starting with Debezium 2.4 the MongoDB connector honors the link:https://www.mongodb.com/docs/manual/core/read-preference/[read preference] configured in either xref:mongodb-property-mongodb-connection-string[`mongodb.connection.string`] or xref:mongodb-property-mongodb-connection-string-shard-params[`mongodb.connection.string.shard.params`] for sharded clusters when xref:mongodb-property-mongodb-connection-mode[`mongodb.connection.mode`] is set to `replica_set`.
```
[the remark this answers is not in the document]
Comment (roldanbob) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: I removed the statement about "Starting with Debezium 2.4", because I  think that it's not necessary. Users who are running earlier versions will read the docs for those versions, and those running 2.4 will read this version. The correct vehicle for alerting user to the change in behavior will be in the 2.4 release notes.

[the remark this answers is not in the document]
Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: I don't know bob. The information about addresses and whatnot is already in the description of `mognodb.connection.string`. What reading preference is is also a common knowledge for anyone using mongo db. So it seems a bit redundant to include it all. I wouldn't mind, however in this form I find the docs "hard to read" -- I just wanted to know how to set the read preference... this is 5+ paragraphs for something which could be 5 sentences (at least that's how I would feel when reading the docs).


Few additional problems

> It then applies the read preferences and other connection parameters that are specified in 
While technically true I think we should  mention that a connection string is constructed and these parameters are used as query parameters. 

> {prodname} tasks use a connection strings that is

Probably should have singular "connection string"

> The connector ignores any read preferences that are specified in the `mongodb.connection.string`. 

That's not true actually... the initial connection (used just to retrieve the topology info) will use it.

[the remark this answers is not in the document]
Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: cc @jpechane  @Naros  What is your opinion, guys? It seems to me that this is getting too complicated. However I might be 100% alone in this feeling.

[the remark this answers is not in the document]
Comment (roldanbob) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: > I don't know bob. The information about addresses and whatnot is already in the description of `mognodb.connection.string`. What reading preference is is also a common knowledge for anyone using mongo db. So it seems a bit redundant to include it all. I wouldn't mind, however in this form I find the docs "hard to read" -- I just wanted to know how to set the read preference... 

OK, @jcechace. I get your point.  I think that I needed to talk my self through the concepts, and I included too much of that in the last version. Let me take one more stab at it to simplify things. Appreciate the input!

[the remark this answers is not in the document]
Comment (Naros) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: @jcechace @roldanbob, so perhaps I'm a bit biased but I have seen lots of users who deploy Debezium who do not have a lot of database experience, and so while specific terminology or constructs may seem commonplace for us often are not for the layperson who is responsible for deploying the Debezium in the runtime environment.  

That said, I do agree that if we get overly verbose in such cases, this can have the opposite effect where users simply refuse to read the documentation because it is too wordy and even more confusing, so striking the right balance is crucial.  I don't know whether this is an opportunity to think about how admonitions can help clarify things for non-native database users in more depth without overly complicating or boring native database users.  Opinions?

[the remark this answers is not in the document]
Comment (roldanbob) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: I agree, @Naros. It's difficult sometimes when I'm trying to describe something I'm unfamiliar with and don't have a good sense of what's common knowledge. 
In my [latest edit](https://github.com/debezium/debezium/pull/4654/files#r1262837732), I think that I moved the needle a bit. Did you guys see that version?

[the remark this answers is not in the document]
Comment (jcechace) [reply documentation/modules/ROOT/pages/connectors/mongodb.adoc:103]: The last iteration looks great, Bob. Thanks!

**Comment by jcechace** [comment]
@roldanbob I've squashed the commit and added all your suggestions. Please give it one final re-read to confirm this is now good for merge :). LGTM

File: documentation/modules/ROOT/pages/connectors/mongodb.adoc [hunk 7]
```diff
@@ -95,6 +95,39 @@ In this way the connector dynamically adjusts to changes in replica set membersh
...
+
+|`replica_set`
+|`mongodb.connection.string.shard.params`
+
+|===
++
+In a sharded cluster, the connector first initiates a connection to the mongos router specified in the `mongodb.connection.string`. 
+For that initial connection, regardless of the connection mode, the connector honors the read preferences that are specified in the `mongodb.connection.string`.
+When the connection mode is set to `replica_set`, after the connector establishes the initial router connection, it retrieves topology information from the router's `config.shards`.
+It then uses the retrieved shard addresses to connect to individual shards in the cluster, constructing connection strings that use the connection parameters in `mongodb.connection.string.shard.params`.
```
Comment (roldanbob) [inline documentation/modules/ROOT/pages/connectors/mongodb.adoc:128]:
```suggestion
It then uses the retrieved shard addresses to connect to individual shards in the cluster, constructing connection strings that use the connection parameters in xref:mongodb-property-mongodb-connection-string-shard-params[`mongodb.connection.string.shard.params`]. 
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
@jcechace I just added one xref, otherwise I think you can merge this. Thanks for your patience in helping me figure out what this was all about.

**Review by roldanbob — APPROVED** [review]
I went ahead and committed the change to add a link to the `mongodb.connection.string.shard.params` property.
